File: the-KINGV2-main/backend/agents/base_agent.py
"""Base Agent Class for LittleHelper AI"""
from abc import ABC, abstractmethod
from typing import Dict, Any, List, Optional, AsyncGenerator
from dataclasses import dataclass, field
import json
import logging

logger = logging.getLogger(__name__)


@dataclass
class AgentResult:
    """Result from an agent execution"""
    success: bool
    content: str
    tokens_used: int = 0
    files_created: List[Dict[str, str]] = field(default_factory=list)  # [{"path": "...", "content": "..."}]
    tasks_generated: List[Dict[str, Any]] = field(default_factory=list)
    errors: List[str] = field(default_factory=list)
    metadata: Dict[str, Any] = field(default_factory=dict)


class BaseAgent(ABC):
    """Abstract base class for all agents"""
    
    agent_id: str = "base"
    agent_name: str = "Base Agent"
    agent_color: str = "#6B7280"
    agent_icon: str = "Bot"
    agent_description: str = "Base agent class"
    
    def __init__(self, ai_service, project_context: Dict[str, Any] = None):
        self.ai_service = ai_service
        self.project_context = project_context or {}
        self.system_prompt = self._build_system_prompt()
    
    @abstractmethod
    def _build_system_prompt(self) -> str:
        """Build the system prompt for this agent"""
        pass
    
    @abstractmethod
    async def execute(self, task: str, context: Dict[str, Any] = None) -> AgentResult:
        """Execute the agent's task"""
        pass
    
    async def execute_streaming(self, task: str, context: Dict[str, Any] = None) -> AsyncGenerator[str, None]:
        """Execute with streaming output"""
        prompt = self._build_prompt(task, context)
        async for chunk in self.ai_service.generate_streaming(prompt, self.system_prompt):
            yield chunk
    
    def _build_prompt(self, task: str, context: Dict[str, Any] = None) -> str:
        """Build the full prompt with context"""
        prompt_parts = []
        
        if self.project_context:
            prompt_parts.append(f"## Project Context")
            prompt_parts.append(f"Language: {self.project_context.get('language', 'Unknown')}")
            prompt_parts.append(f"Project: {self.project_context.get('name', 'Unknown')}")
            if self.project_context.get('description'):
                prompt_parts.append(f"Description: {self.project_context['description']}")
            prompt_parts.append("")
        
        if context:
            if context.get('previous_outputs'):
                prompt_parts.append("## Previous Agent Outputs")
                for output in context['previous_outputs'][-3:]:
                    prompt_parts.append(f"[{output['agent']}]: {output['summary'][:500]}")
                prompt_parts.append("")
            
            if context.get('existing_files'):
                prompt_parts.append("## Existing Files")
                for f in context['existing_files'][:10]:
                    prompt_parts.append(f"- {f['path']}")
                prompt_parts.append("")
            
            if context.get('errors'):
                prompt_parts.append("## Errors to Address")
                for err in context['errors']:
                    prompt_parts.append(f"- {err}")
                prompt_parts.append("")
        
        prompt_parts.append("## Task")
        prompt_parts.append(task)
        
        return "\n".join(prompt_parts)
    
    def _parse_json_from_response(self, response: str) -> Optional[Dict]:
        """Extract JSON from response text"""
        try:
            # Try direct parse first
            return json.loads(response)
        except:
            pass
        
        # Try to find JSON in code blocks
        import re
        json_match = re.search(r'```(?:json)?\s*([\s\S]*?)```', response)
        if json_match:
            try:
                return json.loads(json_match.group(1).strip())
            except:
                pass
        
        # Try to find raw JSON object
        json_match = re.search(r'\{[\s\S]*\}', response)
        if json_match:
            try:
                return json.loads(json_match.group(0))
            except:
                pass
        
        return None
    
    def _extract_code_blocks(self, response: str) -> List[Dict[str, str]]:
        """Extract code blocks from response"""
        import re
        files = []
        
        # Pattern for file path followed by code block
        pattern = r'(?:File:\s*|###\s*|`)?([\w/.-]+\.[\w]+)`?\s*\n```(?:\w+)?\n([\s\S]*?)```'
        matches = re.findall(pattern, response, re.MULTILINE)
        
        for path, content in matches:
            files.append({"path": path.strip(), "content": content.strip()})
        
        return files

File: the-KINGV2-main/backend-csharp/LittleHelperAI.Agents/BaseAgent.cs
// Base Agent Interface and Implementation
using System.Text.Json;
using System.Text.RegularExpressions;

namespace LittleHelperAI.Agents;

/// <summary>
/// Result from an agent execution
/// </summary>
public class AgentResult
{
    public bool Success { get; set; }
    public string Content { get; set; } = "";
    public int TokensUsed { get; set; }
    public List<FileOutput> FilesCreated { get; set; } = new();
    public List<TaskOutput> TasksGenerated { get; set; } = new();
    public List<string> Errors { get; set; } = new();
    public Dictionary<string, object> Metadata { get; set; } = new();
}

public record FileOutput(string Path, string Content);
public record TaskOutput(
    string Id, string Title, string Description, string AgentType,
    int Order, int EstimatedTokens, List<string> Dependencies
);

/// <summary>
/// Interface for all agents
/// </summary>
public interface IAgent
{
    string AgentId { get; }
    string AgentName { get; }
    string AgentColor { get; }
    string AgentIcon { get; }
    string AgentDescription { get; }

    Task<AgentResult> ExecuteAsync(string task, ProjectContext? context = null, ExecutionContext? execContext = null);
    IAsyncEnumerable<string> ExecuteStreamingAsync(string task, ProjectContext? context = null);
}

/// <summary>
/// Project context for agent execution
/// </summary>
public class ProjectContext
{
    public string Name { get; set; } = "";
    public string Language { get; set; } = "Python";
    public string Description { get; set; } = "";
}

/// <summary>
/// Execution context with previous outputs and files
/// </summary>
public class ExecutionContext
{
    public List<PreviousOutput> PreviousOutputs { get; set; } = new();
    public List<ExistingFile> ExistingFiles { get; set; } = new();
    public List<string> Errors { get; set; } = new();
    public string? OriginalRequirements { get; set; }
}

public record PreviousOutput(string Agent, string Summary);
public record ExistingFile(string Path, string Content);

/// <summary>
/// Base implementation for all agents
/// </summary>
public abstract class BaseAgent : IAgent
{
    protected readonly IAIService _aiService;
    
    public abstract string AgentId { get; }
    public abstract string AgentName { get; }
    public abstract string AgentColor { get; }
    public abstract string AgentIcon { get; }
    public abstract string AgentDescription { get; }

    protected BaseAgent(IAIService aiService)
    {
        _aiService = aiService;
    }

    protected abstract string BuildSystemPrompt(ProjectContext? context);

    public abstract Task<AgentResult> ExecuteAsync(string task, ProjectContext? context = null, ExecutionContext? execContext = null);

    public virtual async IAsyncEnumerable<string> ExecuteStreamingAsync(string task, ProjectContext? context = null)
    {
        var prompt = BuildPrompt(task, context, null);
        var systemPrompt = BuildSystemPrompt(context);
        
        await foreach (var chunk in _aiService.GenerateStreamingAsync(prompt, systemPrompt))
        {
            yield return chunk;
        }
    }

    protected string BuildPrompt(string task, ProjectContext? context, ExecutionContext? execContext)
    {
        var parts = new List<string>();

        if (context != null)
        {
            parts.Add("## Project Context");
            parts.Add($"Language: {context.Language}");
            parts.Add($"Project: {context.Name}");
            if (!string.IsNullOrEmpty(context.Description))
                parts.Add($"Description: {context.Description}");
            parts.Add("");
        }

        if (execContext != null)
        {
            if (execContext.PreviousOutputs.Any())
            {
                parts.Add("## Previous Agent Outputs");
                foreach (var output in execContext.PreviousOutputs.TakeLast(3))
                {
                    var summary = output.Summary.Length > 500 ? output.Summary[..500] : output.Summary;
                    parts.Add($"[{output.Agent}]: {summary}");
                }
                parts.Add("");
            }

            if (execContext.ExistingFiles.Any())
            {
                parts.Add("## Existing Files");
                foreach (var file in execContext.ExistingFiles.Take(10))
                {
                    parts.Add($"- {file.Path}");
                }
                parts.Add("");
            }

            if (execContext.Errors.Any())
            {
                parts.Add("## Errors to Address");
                foreach (var error in execContext.Errors)
                {
                    parts.Add($"- {error}");
                }
                parts.Add("");
            }
        }

        parts.Add("## Task");
        parts.Add(task);

        return string.Join("\n", parts);
    }

    protected Dictionary<string, object>? ParseJsonFromResponse(string response)
    {
        try
        {
            return JsonSerializer.Deserialize<Dictionary<string, object>>(response);
        }
        catch { }

        // Try to find JSON in code blocks
        var jsonMatch = Regex.Match(response, @"```(?:json)?\s*([\s\S]*?)```");
        if (jsonMatch.Success)
        {
            try
            {
                return JsonSerializer.Deserialize<Dictionary<string, object>>(jsonMatch.Groups[1].Value.Trim());
            }
            catch { }
        }

        // Try to find raw JSON object
        jsonMatch = Regex.Match(response, @"\{[\s\S]*\}");
        if (jsonMatch.Success)
        {
            try
            {
                return JsonSerializer.Deserialize<Dictionary<string, object>>(jsonMatch.Value);
            }
            catch { }
        }

        return null;
    }

    protected List<FileOutput> ExtractCodeBlocks(string response)
    {
        var files = new List<FileOutput>();
        var pattern = @"###\s*([\w/.\-]+\.[\w]+)\s*\n```(?:[\w]*)?\n([\s\S]*?)```";
        var matches = Regex.Matches(response, pattern);

        foreach (Match match in matches)
        {
            files.Add(new FileOutput(match.Groups[1].Value.Trim(), match.Groups[2].Value.Trim()));
        }

        return files;
    }
}

/// <summary>
/// AI Service interface for agents
/// </summary>
public interface IAIService
{
    Task<AIResponse> GenerateAsync(string prompt, string? systemPrompt = null, int maxTokens = 4000);
    IAsyncEnumerable<string> GenerateStreamingAsync(string prompt, string? systemPrompt = null);
    Task<HealthStatus> CheckHealthAsync();
}

public record AIResponse(string Content, string Provider, string Model, int Tokens);
public record HealthStatus(string Database, string LocalLlm, string Stripe, List<string>? AvailableModels);
